### DISCUSSION TITLE
webpack-dev-server and webpack-dev-middleware crash my app after upgrade to Angular18

### DISCUSSION THREAD

**Discussion by dkoczara** [webpack/webpack#18606] [Q&A]
Hello,
when I upgraded my app to Angular 18 and then the dependencies have been updated:
- webpack-dev-server to version 5.0.4

and

- webpack-dev-middleware to version 7.2.1

my app crushed in dev-mode, because i have some custom webpack logic to add env variables in runetime (only dev, normally app is deployed to nginx, so the problem is only on express dev server).

Example code:

```
			app.use(servePath, (req, res, next) => {
				const originalSend = res.send || res.end;
				res.send = function send(content, ...args) {
                               // mechanism for replacing window.$config in index.html based on mustache and reading values ​​from a json file
					return originalSend.bind(this)(content, ...args);
				};
				next();
			});
```

after update  to Angular 18 the `res.send` method has stopped being called

After hours of debugging i noticed that when I downgraded `webpack-dev-middleware` to version `7.0.0` 

```
package.json:

	"overrides": {
		"@angular-devkit/build-angular": {
			"webpack-dev-server": {
				"webpack-dev-middleware": "7.0.0"
			}
		}
	}
```

then app is working properly, so the problem is in `webpack-dev-middleware`  version `7.1.0` -> at 7.1.0 or higher app start crashing.

Where is the problem ? Is it a `webpack-dev-middleware` or `webpack-dev-server` issue ?

**Comment by alexander-akait** [comment]
You need to implement `stream` support - https://github.com/webpack/webpack-dev-middleware/blob/master/src/utils/compatibleAPI.js#L155

**Reply by dkoczara** [reply]
You mean that the middleware which override `res.send` should return `stream` instead of `return originalSend.bind(this)(content, ...args);` ?

**Reply by alexander-akait** [reply]
No, you need to implement and `stream` and `send`

**Reply by dkoczara** [reply]
Sorry, but I still don't get it. 

In my case at top level I'm using Angular which serve app for development with `@angular-builders/custom-webpack:dev-server` -> which uses under the hood `webpack-dev-server` and `webpack-dev-middleware`. 

My custom implementation of middleware is to override express `res.send` function:

```
				res.send = function send(content, ...args) {
		                        console.log('hello from res.send');
		                        // do my stuff
					return originalSend.bind(this)(content, ...args);
				};
```

and the above code is called and logging only in `webpack-dev-middleware` version less than 7.1.0. 

So in my point of view something is not right, because version 7.1.0 in not a backward compatible (compare to 7.0.0).

However, potential implementations of stream support are transparent from my perspective, because at my level I simply don't see them being anywhere in code

**Reply by alexander-akait** [reply]
We use streams when it is possible (for better memory usage), so your `send` just has not been called

**Comment by awdorrin** [comment]
I see this issue was closed, but this is still an issue that breaks ASP.Net Core/Angular 18 applications that use the legacy SpaProxy for testing in localhost environments. 

@alexander-akait it appears that webpack has introduced an undocumented breaking change that impacts potentially thousands of applications.

**Comment by awdorrin** [comment]
@alexander-akait looks like you fixed this issue with an update to webpack-dev-server v7.4.2 two weeks ago.
Thank you!
